Sources/GeneratorEngine/FileSystem/ReadableFileStream.swift

```diff
@@ -0,0 +1,106 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift open source project
+//
+// Copyright (c) 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SystemPackage
+
+public enum ReadableFileStream: AsyncSequence {
+  public typealias Element = [UInt8]
+
+  case local(LocalReadableFileStream)
+  case virtual(VirtualReadableFileStream)
+
+  public enum Iterator: AsyncIteratorProtocol {
+    case local(LocalReadableFileStream.Iterator)
+    case virtual(VirtualReadableFileStream.Iterator)
+
+    public func next() async throws -> [UInt8]? {
+      switch self {
+      case let .local(local):
+        try await local.next()
+      case let .virtual(virtual):
+        try await virtual.next()
+      }
+    }
+  }
+
+  public func makeAsyncIterator() -> Iterator {
+    switch self {
+    case let .local(local):
+      .local(local.makeAsyncIterator())
+    case let .virtual(virtual):
+      .virtual(virtual.makeAsyncIterator())
+    }
+  }
+}
+
+public struct LocalReadableFileStream: AsyncSequence {
+  public typealias Element = [UInt8]
+
+  let fileDescriptor: FileDescriptor
+  let readChunkSize: Int
+
+  public final class Iterator: AsyncIteratorProtocol {
+    init(_ fileDescriptor: FileDescriptor, readChunkSize: Int) {
+      self.fileDescriptor = fileDescriptor
+      self.readChunkSize = readChunkSize
+    }
+
+    private let fileDescriptor: FileDescriptor
+    private let readChunkSize: Int
+
+    public func next() async throws -> [UInt8]? {
+      var buffer = [UInt8](repeating: 0, count: readChunkSize)
+
+      let bytesRead = try buffer.withUnsafeMutableBytes {
+        try self.fileDescriptor.read(into: $0)
+      }
+
+      guard bytesRead > 0 else {
+        return nil
+      }
+
+      return .init(buffer[0..<bytesRead])
+    }
+
+    deinit {
+      try! fileDescriptor.close()
```

Comment:
this is a very bad idea. Managing file descriptors with `deinit` is all but guaranteed to break everything.

Make the lifecycle explicit or you'll suffer hard. Also `close` should be `async`.

I'd recommend `try await withOpenFile(...) { file in try await file.write(...) }` such that you can't forget the closes.